google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -476,6 +477,29 @@ static String getServiceAccountProjectId(String credentialsPath) {
     return project;
   }
 
+  /* 
+   * Returns true if the projectId is valid. This method checks whether the projectId
+   * contains only lowercase letters, digits and hyphens, starts with a lowercase letter
+   * and does not end with a hyphen, but does not check the length of projectId. This
+   * method is primarily used to protect against DNS hijacking.
+   */
+  static boolean isValidProjectId(String projectId) {   
+    for (char c : projectId.toCharArray()) {
+      if (!isLowerCase(c) && !isDigit(c) && c != '-') {
+        return false;
+      }
+    }
+    return projectId.length() > 0 && isLowerCase(projectId.charAt(0))
```

Comment:
Wouldn't that second condition already be verified by the previous loop?

Reply:
Project-id can only start with a lowercase letter (not a digit or a hyphen). The condition checks this constraint.

Reply:
Oh, so you're using `isLowerCase` to verify that it's a letter. You can see why that's confusing.